### ISSUE TITLE
Some Very Slow CPU Plugins

### ISSUE THREAD

**Issue by ax3l** [ComputationalRadiationPhysics/picongpu#2802]
Some of our regularly used plugins seem to be really, really slow with our `omp2b` backend.

a few that stand out, especially with the very low particle number in my setup:

- PhaseSpace (extremely slow)
- EnergyHistogram (very slow, barely CPU utilization (just 1 per device?) - missing lockstep?)

still verifying:
- PNG
- CountParticles: count macro particles of a species
  (probably same routines used as prep for HDF5/ADIOS output?)
- ADIOS (probably with compression quite expensive? use *more* threads? aggregate?)

OK:
- sumEnergy (fields & particles)
- HDF5 (as usual a bit slow)

## Environment

- Hemera (HZDR) CPUs (24 devices)
- PIConGPU 0.4.1
- box is 1536x2880x1536; cubic cells (dx=dy=dz)
- 3 ion species and 1 electron species (total of ~1.e9 particles; simulation is field dominated)

**Comment by psychocoderHPC** [comment]
Both slow plugins are written in logstep paralism schema else they will not provide correct results. In both cases the reason why it is slow is IMO the global reduce within one device.
It could be that it is faster if we update to the latest alpaka because I implemented native openmp atomic. Atomics between alpaka blocks are in alpaka 0.3.X OpenMP barriers which are very slow.

**Comment by ax3l** [comment]
Yep, might be. Interesting though, the global reduces such as "count" (0D reduce) are still ok. Just the phase-space (2D map-reduce) and energy histogram (1D map-reduce) are really slow.

**Comment by psychocoderHPC** [comment]
Accross blocks within a grid.
But please do not associate global memory with accross block atomics because you can use atomics those are only save between threads in a alpaka block on global memory. This means alpaka atomics are memory hierarchy independent.

**Comment by ax3l** [comment]
Yes, that makes sense. I just wanted to put in brackets what we are doing in those two plugins currently on GPU.

**Comment by ax3l** [comment]
After about 9hrs with no initial output, I can't say for sure if it's not also deadlocking (with high CPU load) in some cases for those plugins or if the barriers are just extraordinary slow currently. Nevertheless, it does not seem energy histograms and phase space can be used in CPU production yet due to it.

**Comment by ax3l** [comment]
cc @sbastrakov @tdd11235813 this is the Alpaka 3.4 OpenMP 2.0 (blocks) map-reduce performance issue we currently face in PIConGPU.

**Comment by ax3l** [comment]
@psychocoderHPC I update my local alpaka on PIConGPU 0.4.1 to https://github.com/ComputationalRadiationPhysics/alpaka/pull/698 and it improved the situation significantly. (Using GCC 7.30 on Hemera with OpenMP 2 blocks backend but OpenMP 4.5 available.)

Thanks a ton, this is awesome! :sparkles: :sparkling_heart:
